### DISCUSSION TITLE
emscripten support?

### DISCUSSION THREAD

**Discussion by esotericist** [dankamongmen/notcurses#2692] [Q&A]
i haven't been in a position to push further with my experiments with either notcurses or emscripten, so if it turns out this is already supported and somehow there's simply nothing in the issues, pull requests, discussions, or repo about it, i apologize in advance. 

i'm still plotting some of the stuff for my app, but "as much cross platform as i can squeeze out of not having to bother with platform-specific code" is one of my goals, and having a competent tui that just works on a few platforms would really help a bunch. right now i'm more or less waffling between notcurses, which is awesome, and imtui, which looks a fair bit more approachable in some ways, and explicitly has some emscripten support, but lacks a lot of things notcurses can do (including more competent windows terminal support).

what would go into using notcurses to work for a web deployment?

**Comment by dankamongmen** [comment]
so first off, i know essentially nothing about web development--i'm not even certain what "emscripten" is. its own language? an umbrella term for javascript and similar technologies? when this mail hit my box, i thought this was going to be an emacs question, lol. 

...looks it up..."Emscripten is a complete compiler toolchain to WebAssembly, using LLVM" ok, well that sounds reasonable enough; i know all those words. 

...looks up imtui...stops reading at "256 ANSI colors" alright so i've done no deep dive into imtui, and i'm sure the authors are fine people, but this looks like yet another early-aughts-style NCURSES alternative.

so i can't answer your question about which would be better for web deployment, as i'm simply too ignorant to do so. if imlib can be made to run in a web browser, i'm sure notcurses can, as well (one or both might require some small changes, but i'd be happy to do them if brought to my attention). if you're able to supply a Terminfo database and render terminfo-augmented lexemes to a web page, notcurses can roll, and you'd surely need these two capabilities for any such library to operate.

as for notcurses vs imlib: i'm something of a maximalist developer, yet fanatical about minutiae, and when writing a library also generalism in my APIs. i'm very hesitant to do special cases and one-offs, *even if they might be the most approachable API for a new user*, if there's a reasonable and performant generalization. maximalism means i'm going to try and discover the entire space (as a result, Notcurses development led to bugs being found and fixed in *every* terminal I've tested it against), and  covering a large, diverse space requires fairly abstract generalizations. at the same time, facilitating peak performance requires specializations, departures from the general.

in designing Notcurses, i strove to be general, even at the expense of simplicity, except for when performance mandated otherwise--and performance is never simple. i agree that it is intimidating on first approach (a fact not helped by my aggressively formal and axiomatic documentation). most libraries lead with friendly "getting started" guides (something @joseluis  has requested from me more than once); i point you at page-length comments in `notcurses.h` full of unicode and suggest you read [the book](https://nick-black.com/htp-notcurses.pdf). i could probably do a lot better here, and apologize. with that said, once you grok the central data structures (`ncplane`, `ncvisual`, and friends), i've gone to great effort to build these powerful generalizations around sensible defaults. you can construct tremendous things with very little notcurses code. take a look through `src/demo` -- there are any number of results possible in no other existing library, or capable of being expressed there only grotesquely.

tl;dr: no answer to your question, but i'm the best [citation needed]

hack on, and i hope notcurses proves useful to you!

**Reply by esotericist** [reply]
as i understand it, imtui is *not* trying to be a general alternative to ncurses, it is actually using ncurses as its backend (even for web deployments, as far as i can tell), it's simply providing widget frameworks based off of imgui's api. and generally it looks like it's trying to be a more approachable answer to the much more venerable semantics.

re: documentation, the .h and The Book are things i've looked at, and... yeah. that's a steep hill to climb for me, especially since i'm not actually all that comfortable with c, i just found myself restricted to it by running out of other options for my objectives (\*\*). but "this is exceedingly difficult" is not an impassable wall in and of itself, nor was it the direct concern i had at hand; it's a question of "how much labor will it take to get the support matrix covered adequately".

notably: imtui's windows support is visibly behind yours, which will probably alarm you given what i've seen in your issues in that space. so just on that front, notcurses is already ahead of the game for my purposes, nevermind the fact that many of the bling-as-hell features of notcurses are slick and desirable. so unless i ultimately decide i genuinely can't comprehend your api, i'm likely to stick to notcurses just for those two factors.

but now that i know "a tui on a web browser is apparently a lot more reasonable than i thought", i figured i'd come check if anyone else had, at some point, had the thought of "what if we could run dank's mad demos straight on the web. and also maybe actually useful apps."

> if you're able to supply a Terminfo database and render terminfo-augmented lexemes to a web page, notcurses can roll, and you'd surely need these two capabilities for any such library to operate.

i recognize most of those words. i don't think i have a solid enough grasp to provide what you're asking for, though. which is probably fine since i'm not in a position to actually be looking to deploy anything anyway; this is the cart on the road sans horses. they're still back at the stable.

beyond that: thanks for the reply, though it takes me a few passes to work fully through. i am delighted by your stream of consciousness communications, but they aren't always the easiest for me to process with my cognitive restrictions XD

_** my specific use case is trying to build a managed environment that suits my cognitive needs which is most emphatically not anything like a c-heritage language; i've just concluded pure c is the least painful way to get the vm working how i need it, within the limits of my cognition. which confuses and distresses a lot of programmers i occasionally ask for help from, because they generally tend to think i should be using something else. or think i should not be trying to build a managed stack-based concurrency focused environment in the first place. which is understandable, because my objective is pretty arcane on the surface._

**Reply by esotericist** [reply]
in terrible ideas i hope to convince myself not to follow up on:

one ridiculous option is to use dear imgui (the upstream graphical ui library that imtui is providing an api mirrored from), and use notcurses as a blitting backend instead of e.g. sdl2.

this is obviously a losing proposition from a complexity standpoint alone. and yet, there's a part of my brain that goes "but eso! mssr dank assures us that notcurses is fast as shit!". 

hopefully, i'll be able to find a suitable dark alley to drag this notion into in order to beat it until it learns not to come back.

**Reply by andreamancuso** [reply]
I've been trying to write a basic Dear ImGui backend using notcurses blitting - unsuccessfully so far. Wondering if anyone is interested in joining forces?

**Reply by esotericist** [reply]
> I've been trying to write a basic Dear ImGui backend using notcurses blitting - unsuccessfully so far. Wondering if anyone is interested in joining forces?

i wish i could say yes, but i sadly can't. i do wish you the best of luck and science, though

**Reply by andreamancuso** [reply]
Thank you, and fair enough.

I'm using Imtui for now.
